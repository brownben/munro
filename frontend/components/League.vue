<script setup lang="ts">
import type { PropType } from 'vue'
import type { League } from '~/api-types'

defineProps({
  league: { type: Object as PropType<League>, required: true },
  h2: { type: Boolean, default: true },
})
</script>

<template>
  <article>
    <h2
      v-if="h2"
      class="text-3xl font-black text-gray-900 sm:text-4xl dark:text-gray-100"
    >
      {{ league.name }}
    </h2>
    <h3
      v-else
      class="text-3xl font-black text-gray-900 sm:text-4xl dark:text-gray-100"
    >
      {{ league.name }}
    </h3>

    <p class="pt-2 leading-tight text-gray-600 sm:text-xl dark:text-gray-400">
      {{ league.tagline }}
    </p>

    <NuxtLink
      :to="`/leagues/${league.name}`"
      class="mt-6 inline-block text-main-700 hover:text-main-600 dark:text-main-600 dark:hover:text-main-500"
    >
      View details â†’
    </NuxtLink>
  </article>
</template>
