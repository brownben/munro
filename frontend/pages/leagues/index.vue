<script setup lang="ts">
import type { League } from '~/api-types'

const { data: leagues } = await useData<League[]>('leagues/')

useTitle({
  title: 'Leagues',
  description: 'All the leagues with results on Munro',
})
</script>

<template>
  <div>
    <Heading title="Leagues" />
    <main
      class="mx-auto max-w-screen-lg gap-8 px-6 py-10 sm:py-12 md:py-14 lg:px-8"
    >
      <div
        class="col-span-2 -mx-4 -mb-4 -mt-8 flex flex-col divide-y dark:divide-gray-600"
      >
        <League
          v-for="league in leagues"
          :key="league.name"
          :league="league"
          class="px-4 py-10"
        />
      </div>
    </main>
  </div>
</template>
