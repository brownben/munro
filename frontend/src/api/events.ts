import { getData, postData, putData, deleteData } from './requests'

export const getEvent = (id: string): Promise<LeagueEvent | null> =>
  getData<LeagueEvent>({
    apiLocation: `/api/events/${id}`,
    customErrorMessage: 'Problem Fetching Event Details',
  })

export const getEvents = (): Promise<LeagueEvent[] | null> =>
  getData<LeagueEvent[]>({
    apiLocation: '/api/events',
    customErrorMessage: 'Problem Fetching Events',
  })

export const getLeagueEvents = (
  league: string,
  uploadKey?: boolean
): Promise<LeagueEvent[] | null> => {
  const apiLocation = uploadKey
    ? `/api/leagues/${league}/events/uploadKey`
    : `/api/leagues/${league}/events`

  return getData<LeagueEvent[]>({
    apiLocation,
    customErrorMessage: 'Problem Fetching Events',
  })
}

export const getLatestResults = (): Promise<LeagueEvent[] | null> =>
  getData<LeagueEvent[]>({
    apiLocation: '/api/events/latest-results',
    customErrorMessage: 'Problem Fetching Events',
  })

export const createEvent = (data: LeagueEvent): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/events`,
    data,
    customErrorMessage: 'Problem Creating Event',
    customSuccessMessage: `Event \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateEvent = (data: LeagueEvent): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/events/${data.id}`,
    data,
    customErrorMessage: 'Problem Updating Event',
    customSuccessMessage: `Event \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteEvent = (
  id: string,
  name: string
): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/events/${id}`,
    customErrorMessage: 'Problem Deleting Event',
    customSuccessMessage: `Event \`${name}\` Deleted`,
  })
