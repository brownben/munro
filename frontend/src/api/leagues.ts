import { getData, postData, putData, deleteData } from './requests'

export const getLeague = (name: string): Promise<League | null> =>
  getData<League | null>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Fetching League Details',
  })

export const getLeagues = (): Promise<League[] | null> =>
  getData<League[]>({
    apiLocation: '/api/leagues',
    customErrorMessage: 'Problem Fetching Leagues',
  })

export const createLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  postData<ServerMessage>({
    apiLocation: `/api/leagues`,
    data,
    customErrorMessage: 'Problem Creating League',
    customSuccessMessage: `League \`${data.name}\` Created`,
    customErrorHandler: true,
  })

export const updateLeague = (data: LeagueForm): Promise<ServerMessage | null> =>
  putData<ServerMessage>({
    apiLocation: `/api/leagues/${data.name}`,
    data,
    customErrorMessage: 'Problem Updating League',
    customSuccessMessage: `League \`${data.name}\` Updated`,
    customErrorHandler: true,
  })

export const deleteLeague = (name: string): Promise<ServerMessage | null> =>
  deleteData<ServerMessage>({
    apiLocation: `/api/leagues/${name}`,
    customErrorMessage: 'Problem Deleting League',
    customErrorHandler: true,
  })
