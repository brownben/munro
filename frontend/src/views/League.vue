<!--
  League Home Page

  Shows all league details as well as all details for each event in the league, including
  the upload key and event id needed for eveent upload if logged in. If logged in it also diaplays
  options to edit/ update/ delete the events/ league. Also has links to results for each course
-->

<template>
  <div>
    <div v-if="league">
      <div id="league-header">
        <h1>{{ league.name }}</h1>
        <p v-if="league.moreInformation">{{ league.moreInformation }}</p>
        <p
          v-if="league.courses"
        >There are normally {{ league.courses.length }} courses - {{ league.courses.join(', ') }}</p>
        <p v-if="league.coordinator">{{ league.coordinator }} coordinates the league.</p>
        <p>
          <span
            v-if="league.scoringMethod"
          >The scoring for the league is calculated using a {{ league.scoringMethod }} based method.</span>
          <span
            v-if="league.numberOfCountingEvents && league.numberOfEvents"
          >With your best {{ league.numberOfCountingEvents }} events from all {{ league.numberOfEvents }} events counting.</span>
        </p>
        <p v-if="league.website">
          More information can be found at
          <a :href="league.website">{{ league.website }}</a>
        </p>
      </div>
      <div v-if="auth.isLoggedIn" class="actions">
        <button @click="$router.push($route.path+'/create-event')">Add Event</button>
        <button @click="$router.push($route.path+'/edit')">Edit League</button>
        <button @click="deleteLeague()">Delete League</button>
      </div>
      <div class="actions">
        <button
          v-for="course of league.courses"
          :key="course"
          @click="$router.push($route.path + '/' + course)"
        >{{ course }}</button>
      </div>
      <div class="events">
        <div v-for="event of events" :key="event.name" class="event">
          <h2>{{ event.name }}</h2>
          <div v-if="auth.isLoggedIn" class="event-actions">
            <button @click="$router.push('/events/'+event.id+'/edit')">Edit Event</button>
            <button @click="$router.push('/upload/'+event.id)">Upload Results</button>
            <button @click="deleteEvent(event)">Delete Event</button>
          </div>
          <p v-if="auth.isLoggedIn">
            <b>Event ID:</b>
            {{ event.id }}
          </p>
          <p v-if="auth.isLoggedIn && event.uploadKey">
            <b>Event Upload Key:</b>
            {{ event.uploadKey }}
          </p>
          <p>On {{ event.date.split('-')[2] }}/{{ event.date.split('-')[1] }}/{{ event.date.split('-')[0] }} organised by {{ event.organiser }}</p>
          <p v-if="event.moreInformation">{{ event.moreInformation }}</p>
          <p v-if="event.website">
            More Information can be found at
            <a :href="event.website">{{ event.website }}</a>
          </p>
          <div v-if="event.resultUploaded" class="event-actions">
            <a v-if="event.results" :href="event.results" class="button">Results</a>
            <a v-if="event.winsplits" :href="event.winsplits" class="button">WinSplits</a>
            <a v-if="event.routegadget" :href="event.routegadget" class="button">Routegadget</a>
          </div>
        </div>
      </div>
    </div>
    <not-found v-if="!league" />
  </div>
</template>
<script>
import axios from 'axios'
import NotFound from '@/views/NotFound'

export default {
  components: {
    'NotFound': NotFound,
  },

  data: function () {
    return {
      league: {},
      events: [],
      auth: this.$auth,
    }
  },

  watch: {
    // Update details if the league in the URL changes (VueJS problem where no reload if the parameter part changes, so needs watched)
    '$route': function () {
      this.getLeague()
        .then(() => this.getLeagueEvents())
    },
  },

  mounted: function () {
    // Get details on load
    this.getLeague()
      .then(() => this.getLeagueEvents())
  },

  methods: {
    getLeague: function () {
      return axios.get('/api/leagues/' + this.$route.params.name)
        .then(response => { this.league = response.data })
        .catch(() => this.$messages.addMessage('Problem Getting League Details'))
    },

    getLeagueEvents: function () {
      if (this.league) {
        if (this.auth.isLoggedIn) {
          axios.get('/api/leagues/' + this.league.name + '/events/uploadKey')
            .then(response => { this.events = response.data })
            .catch(() => this.$messages.addMessage('Problem Getting Event Details'))
        }
        else {
          axios.get('/api/leagues/' + this.league.name + '/events')
            .then(response => { this.events = response.data })
            .catch(() => this.$messages.addMessage('Problem Getting Event Details'))
        }
      }
    },

    deleteLeague: function () {
      if (confirm('Are you Sure you Want to Delete League - ' + this.league.name + '? \nThis Action Can\'t Be Recovered')) {
        axios.delete('/api/leagues/' + this.league.name)
          .then(() => {
            this.$messages.addMessage('League: ' + this.league.name + ' was Deleted')
            this.$router.push('/')
          })
          .catch(() => this.$messages.addMessage('Problem Deleting League - Please Try Again'))
      }
    },

    deleteEvent: function (event) {
      if (confirm('Are you Sure you Want to Delete Event - ' + event.name + '? \nThis Action Can\'t Be Recovered')) {
        axios.delete('/api/events/' + event.id)
          .then(() => {
            this.$messages.addMessage('Event: ' + event.name + ' was Deleted')
            this.getLeague()
              .then(() => this.getLeagueEvents())
          })
          .catch(() => this.$messages.addMessage('Problem Deleting Event - Please Try Again'))
      }
    },
  },
}
</script>
<style lang="stylus" scoped>
@import '../assets/styles/helpers.styl'
@import '../assets/styles/inputs.styl'

#league-header
  margin-bottom: 1rem
  padding: 0.5rem 5% 1rem
  background-color: purple-400
  color: white
  box-shadow(1.5)

  h1
    padding: 0.5rem 0
    color: white
    font-size: 2.5rem

  p, a
    color: white

  @media (max-width: 700px)
    margin-bottom: 1.5rem
    padding: 1rem
    padding-top: 0.5rem

.actions
  margin-left: 5%
  width: 90%

  button
    border: 0
    box-shadow(1)
    margin-right: 1rem
    margin-bottom: 0.75rem
    padding: 0.5rem 1.5rem
    border: 0

    &:hover
      background-color: white
      color: purple-500
      box-shadow(2)

  @media (max-width: 700px)
    button
      margin-top: 0.5rem
      margin-left: 0
      width: 100%

.event
  box-sizing: border-box
  margin-top: 1.5rem
  margin-left: 5%
  padding: 0.75rem
  width: 90%
  box-shadow(1)

  &:first-child
    margin-top: 0.75rem

  h2
    padding: 0.25rem 0

  p
    padding: 0.15rem 0

  .event-actions
    padding: 0.25rem 0 0

    button, .button
      margin-bottom: 0.25rem

    .button
      display: inline-block
      box-sizing: border-box

    a
      text-decoration: none
</style>
