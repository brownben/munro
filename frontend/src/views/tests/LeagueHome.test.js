/*
  League Home View Unit Tests
*/

import axios from 'axios'
import VueRouter from 'vue-router'
import { mount, shallowMount, createLocalVue } from '@vue/test-utils'

import League from '@/views/LeagueHome'
import { sampleSingleLeague } from '@/tests/test data/leagues'
import { sampleSingleEvent, sampleThreeEvents } from '@/tests/test data/events'

jest.mock('axios')
const flushPromises = () => new Promise(resolve => global.setImmediate(resolve))

beforeEach(() => {
  jest.resetAllMocks()
  axios.get.mockRejectedValue()
})

test('Is a Vue Instance', () => {
  const mockGetLeague = jest.fn().mockResolvedValue()
  const mockGetLeagueEvents = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
    },
    methods: {
      getLeague: mockGetLeague,
      getLeagueEvents: mockGetLeagueEvents,
    },
    stubs: ['router-link', 'vue-headful'],
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const mockGetLeague = jest.fn().mockResolvedValue()
  const mockGetLeagueEvents = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
    },
    methods: {
      getLeague: mockGetLeague,
      getLeagueEvents: mockGetLeagueEvents,
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ league: sampleSingleLeague, events: sampleThreeEvents })
  expect(wrapper.element).toMatchSnapshot()
})

test('Calls Correct Functions on Load', async () => {
  const mockGetLeague = jest.fn().mockResolvedValue()
  const mockGetLeagueEvents = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
    },
    methods: {
      getLeague: mockGetLeague,
      getLeagueEvents: mockGetLeagueEvents,
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  await wrapper.vm.$mount()
  expect(mockGetLeague).toHaveBeenCalledTimes(1)
  expect(mockGetLeague).toHaveBeenLastCalledWith()
  expect(mockGetLeagueEvents).toHaveBeenCalledTimes(1)
  expect(mockGetLeagueEvents).toHaveBeenLastCalledWith()
})

test('Reload League on Route Change', async () => {
  const mockGetLeague = jest.fn().mockResolvedValue()
  const mockGetLeagueEvents = jest.fn()
  const localVue = createLocalVue()
  localVue.use(VueRouter)
  const router = new VueRouter({ routes: [{ path: '/' }, { path: '/other' }] })
  shallowMount(League, {
    localVue,
    router,
    mocks: {
      $auth: { user: false },
    },
    methods: {
      getLeague: mockGetLeague,
      getLeagueEvents: mockGetLeagueEvents,
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  expect(mockGetLeague).toHaveBeenCalledTimes(0)
  expect(mockGetLeagueEvents).toHaveBeenCalledTimes(0)
  router.push('/other')
  await flushPromises()
  expect(mockGetLeague).toHaveBeenCalledTimes(1)
  expect(mockGetLeague).toHaveBeenLastCalledWith()
  expect(mockGetLeagueEvents).toHaveBeenCalledTimes(1)
  expect(mockGetLeagueEvents).toHaveBeenLastCalledWith()
})

test('Get League - Correct API Call', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: 'Test' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue()
  await wrapper.vm.getLeague()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/leagues/Test')
})

test('Get League - Success', async () => {
  const mockAddMessage = jest.fn()
  axios.get.mockResolvedValue({ data: sampleSingleLeague[0] })
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await wrapper.vm.getLeague()
  expect(mockAddMessage).toHaveBeenCalledTimes(0)
  expect(wrapper.vm.league).toEqual(sampleSingleLeague[0])
})

test('Get League - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  await wrapper.vm.getLeague()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Problem Getting League Details')
})

test('Get League Events - No League Data', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue()
  wrapper.setData({ league: false })
  expect(wrapper.vm.getLeagueEvents()).toBeFalsy()
  expect(axios.get).toHaveBeenCalledTimes(0)
  expect(mockAddMessage).toHaveBeenCalledTimes(0)
})

test('Get League Events - Correct API', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue()
  wrapper.setData({ league: { name: 'Test' } })
  await wrapper.vm.getLeagueEvents()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/leagues/Test/events')
})

test('Get League Events - Correct API - Logged In', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: true },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue()
  wrapper.setData({ league: { name: 'Test' } })
  await wrapper.vm.getLeagueEvents()
  expect(axios.get).toHaveBeenCalledTimes(1)
  expect(axios.get).toHaveBeenLastCalledWith('/api/leagues/Test/events/uploadKey')
})

test('Get League Events - Success', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: sampleSingleEvent })
  wrapper.setData({ league: sampleSingleLeague[0] })
  await wrapper.vm.getLeagueEvents()
  expect(wrapper.vm.events).toEqual(sampleSingleEvent)
})

test('Get League Events - Success - Logged In', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: true },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockResolvedValue({ data: sampleSingleEvent })
  wrapper.setData({ league: sampleSingleLeague[0] })
  await wrapper.vm.getLeagueEvents()
  expect(wrapper.vm.events).toEqual(sampleSingleEvent)
})

test('Get League Events - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  wrapper.setData({ league: { name: 'Test' } })
  await wrapper.vm.getLeagueEvents()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Problem Getting Event Details')
})

test('Get League Events - Error - Logged In', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: true },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  axios.get.mockRejectedValue()
  wrapper.setData({ league: { name: 'Test' } })
  await wrapper.vm.getLeagueEvents()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Problem Getting Event Details')
})

test('Delete Event - Correct Confirm', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  global.confirm = jest.fn()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(global.confirm).toHaveBeenCalledTimes(1)
  expect(global.confirm).toHaveBeenLastCalledWith('Are you Sure you Want to Delete Event - EVENT? \nThis Action Can\'t Be Recovered')
})

test('Delete Event - Denied Confirmation', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  global.confirm = jest.fn().mockReturnValue(false)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(axios.delete).toHaveBeenCalledTimes(0)
})

test('Delete Event - Correct API Call', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(axios.delete).toHaveBeenCalledTimes(1)
  expect(axios.delete).toHaveBeenLastCalledWith('/api/events/EVENT')
})

test('Delete Event - Success', async () => {
  const mockGetLeague = jest.fn().mockResolvedValue()
  const mockGetLeagueEvents = jest.fn()
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
    methods: {
      getLeague: mockGetLeague,
      getLeagueEvents: mockGetLeagueEvents,
    },
  })
  await flushPromises()
  jest.clearAllMocks()
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Event: EVENT was Deleted')
  expect(mockGetLeague).toHaveBeenCalledTimes(1)
  expect(mockGetLeagueEvents).toHaveBeenCalledTimes(1)
})

test('Delete Event - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockRejectedValue()
  await wrapper.vm.deleteEvent({ id: 'EVENT', name: 'EVENT' })
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Problem Deleting Event - Please Try Again')
})

test('Delete League - Correct Confirm', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  global.confirm = jest.fn()
  wrapper.setData({ league: { name: 'LEAGUE' } })
  await wrapper.vm.deleteLeague()
  expect(global.confirm).toHaveBeenCalledTimes(1)
  expect(global.confirm).toHaveBeenLastCalledWith('Are you Sure you Want to Delete League - LEAGUE? \nThis Action Can\'t Be Recovered')
})

test('Delete League - Denied Confirmation', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ league: { name: 'LEAGUE' } })
  global.confirm = jest.fn().mockReturnValue(false)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteLeague()
  expect(axios.delete).toHaveBeenCalledTimes(0)
})

test('Delete League - Correct API Call', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ league: { name: 'LEAGUE' } })
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteLeague()
  expect(axios.delete).toHaveBeenCalledTimes(1)
  expect(axios.delete).toHaveBeenLastCalledWith('/api/leagues/LEAGUE')
})

test('Delete League - Success', async () => {
  const mockGetLeague = jest.fn().mockResolvedValue()
  const mockGetLeagueEvents = jest.fn()
  const mockAddMessage = jest.fn()
  const mockRouterPush = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $router: { push: mockRouterPush },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    methods: {
      getLeague: mockGetLeague,
      getLeagueEvents: mockGetLeagueEvents,
    },
    stubs: ['router-link', 'vue-headful'],
  })
  wrapper.setData({ league: { name: 'LEAGUE' } })
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockResolvedValue()
  await wrapper.vm.deleteLeague()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('League: LEAGUE was Deleted')
  expect(mockRouterPush).toHaveBeenCalledTimes(1)
  expect(mockRouterPush).toHaveBeenLastCalledWith('/')
})

test('Delete League - Error', async () => {
  const mockAddMessage = jest.fn()
  const wrapper = shallowMount(League, {
    mocks: {
      $auth: { user: false },
      $route: { params: { name: '' } },
      $messages: { addMessage: mockAddMessage },
      mounted: () => { },
    },
    stubs: ['router-link', 'vue-headful'],
  })
  await flushPromises()
  jest.clearAllMocks()
  wrapper.setData({ league: { name: 'LEAGUE' } })
  global.confirm = jest.fn().mockReturnValue(true)
  axios.delete.mockRejectedValue()
  await wrapper.vm.deleteLeague()
  expect(mockAddMessage).toHaveBeenCalledTimes(1)
  expect(mockAddMessage).toHaveBeenLastCalledWith('Problem Deleting League - Please Try Again')
})

test('Scoring Method Shorthand to Full', () => {
  const wrapper = mount(League, {
    mocks: {
      $route: { path: '', params: { name: '' } },
      $messages: { addMessage: jest.fn() },
    },
    stubs: ['dropdown-input', 'router-link', 'vue-headful'],
  })
  expect(wrapper.vm.scoringMethodShorthandToFull('position')).toBe('Position Based (100 Max)')
  expect(wrapper.vm.scoringMethodShorthandToFull('position50')).toBe('Position Based (50 Max)')
  expect(wrapper.vm.scoringMethodShorthandToFull('position99')).toBe('Position Based (99 Max)')
  expect(wrapper.vm.scoringMethodShorthandToFull('position99average')).toBe('Position Based (99 Max, Reduced in a Draw)')
  expect(wrapper.vm.scoringMethodShorthandToFull('positionDouble')).toBe('Position Based (100 Max, Double Points)')
  expect(wrapper.vm.scoringMethodShorthandToFull('position50Double')).toBe('Position Based (50 Max, Double Points)')
  expect(wrapper.vm.scoringMethodShorthandToFull('timeAverage')).toBe('Relative to Average Time (1000 Average)')
  expect(wrapper.vm.scoringMethodShorthandToFull('timeAverage100')).toBe('Relative to Average Time (100 Average)')
  expect(wrapper.vm.scoringMethodShorthandToFull('')).toBe('')
  expect(wrapper.vm.scoringMethodShorthandToFull('pos')).toBe('')
})
