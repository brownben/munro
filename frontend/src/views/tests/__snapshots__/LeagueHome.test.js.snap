// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<div>
  <vue-headful-stub
    description="Event Information and Results for the
         league on Munro, the Fast and Easy Results System for Orienteering Leagues. A simple way to calculate the results for orienteering leagues, with search and sort features"
    head="[object Object]"
    title="Munro - "
    url="https://munro-leagues.herokuapp.com/leagues/"
  />
   
  <!---->
   
  <div
    class="w-full bg-main text-white p-3 text-center"
  >
    <h2
      class="text-2xl font-heading"
    >
      League Results
    </h2>
     
    <div
      class="pb-1"
    />
  </div>
   
  <div
    class="view"
  >
    <!---->
     
    <div
      class="card my-4"
    >
      <h2
        class="font-heading text-xl my-1"
      >
        
        200m
      
      </h2>
       
      <!---->
       
      <div
        class="my-1"
      >
        <!---->
         
        <!---->
      </div>
       
      <!---->
       
      <!---->
       
      <p>
        
        More Information can be found on their
        
        <a
          class="link"
          href="wwwwwww"
          rel="noopener noreferrer"
          target="_blank"
        >
          website
        </a>
      </p>
       
      <!---->
    </div>
    <div
      class="card my-4"
    >
      <h2
        class="font-heading text-xl my-1"
      >
        
        Longniddry
      
      </h2>
       
      <!---->
       
      <div
        class="my-1"
      >
        <!---->
         
        <!---->
      </div>
       
      <p>
        
        On 18/04/2018
        organised by ELO
      </p>
       
      <!---->
       
      <!---->
       
      <!---->
    </div>
    <div
      class="card my-4"
    >
      <h2
        class="font-heading text-xl my-1"
      >
        
        South Edinburgh
      
      </h2>
       
      <!---->
       
      <div
        class="my-1"
      >
        <!---->
         
        <!---->
      </div>
       
      <p>
        
        On 25/04/2018
        organised by INT
      </p>
       
      <!---->
       
      <!---->
       
      <!---->
    </div>
  </div>
   
  <!---->
</div>
`;
